File: java/social-media-photo-app/src/test/java/com/rbleggi/socialmedia/SocialMediaTest.java
package com.rbleggi.socialmedia;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class SocialMediaTest {
    private User user;
    private TestFollower follower1;
    private TestFollower follower2;

    @BeforeEach
    void setUp() {
        user = new User("Alice");
        follower1 = new TestFollower("Bob");
        follower2 = new TestFollower("Charlie");
    }

    @Test
    @DisplayName("User should notify followers when photo is published")
    void user_publishPhoto_notifiesFollowers() {
        user.registerObserver(follower1);
        user.registerObserver(follower2);

        var photo = new Photo("1", "http://photo.com/1", List.of());
        user.publishPhoto(photo);

        assertEquals(1, follower1.events.size());
        assertEquals(1, follower2.events.size());
        assertInstanceOf(PhotoPublished.class, follower1.events.get(0));
    }

    @Test
    @DisplayName("User should notify followers when user is tagged")
    void user_tagUser_notifiesFollowers() {
        user.registerObserver(follower1);

        var photo = new Photo("1", "http://photo.com/1", List.of());
        user.tagUser(photo, "Charlie");

        assertEquals(1, follower1.events.size());
        assertInstanceOf(UserTagged.class, follower1.events.get(0));
        assertEquals("Charlie", ((UserTagged) follower1.events.get(0)).taggedUser());
    }

    @Test
    @DisplayName("User should notify followers when photo is commented")
    void user_commentPhoto_notifiesFollowers() {
        user.registerObserver(follower1);

        var photo = new Photo("1", "http://photo.com/1", List.of());
        var comment = new Comment("Bob", "Great photo!");
        user.commentPhoto(photo, comment);

        assertEquals(1, follower1.events.size());
        assertInstanceOf(PhotoCommented.class, follower1.events.get(0));
    }

    @Test
    @DisplayName("User should be able to register multiple observers")
    void user_registerMultipleObservers_registersAll() {
        user.registerObserver(follower1);
        user.registerObserver(follower2);

        var photo = new Photo("1", "http://photo.com/1", List.of());
        user.publishPhoto(photo);

        assertEquals(1, follower1.events.size());
        assertEquals(1, follower2.events.size());
    }

    @Test
    @DisplayName("User should be able to remove observer")
    void user_removeObserver_stopsNotifications() {
        user.registerObserver(follower1);
        user.removeObserver(follower1);

        var photo = new Photo("1", "http://photo.com/1", List.of());
        user.publishPhoto(photo);

        assertEquals(0, follower1.events.size());
    }

    @Test
    @DisplayName("Follower should handle PhotoPublished event")
    void follower_photoPublished_handlesEvent() {
        var follower = new Follower("TestFollower");
        var photo = new Photo("1", "http://photo.com/1", List.of());
        var event = new PhotoPublished(photo);
        assertDoesNotThrow(() -> follower.update(event));
    }

    @Test
    @DisplayName("Follower should handle UserTagged event")
    void follower_userTagged_handlesEvent() {
        var follower = new Follower("TestFollower");
        var photo = new Photo("1", "http://photo.com/1", List.of());
        var event = new UserTagged(photo, "SomeUser");
        assertDoesNotThrow(() -> follower.update(event));
    }

    @Test
    @DisplayName("Follower should handle PhotoCommented event")
    void follower_photoCommented_handlesEvent() {
        var follower = new Follower("TestFollower");
        var photo = new Photo("1", "http://photo.com/1", List.of());
        var comment = new Comment("User", "Nice!");
        var event = new PhotoCommented(photo, comment);
        assertDoesNotThrow(() -> follower.update(event));
    }

    @Test
    @DisplayName("Photo should hold correct data")
    void photo_holdsCorrectData() {
        var photo = new Photo("1", "http://photo.com/1", List.of("tag1", "tag2"));
        assertEquals("1", photo.id());
        assertEquals("http://photo.com/1", photo.url());
        assertEquals(2, photo.tags().size());
    }

    @Test
    @DisplayName("Comment should hold correct data")
    void comment_holdsCorrectData() {
        var comment = new Comment("User", "Great photo!");
        assertEquals("User", comment.user());
        assertEquals("Great photo!", comment.text());
    }

    private static class TestFollower implements Observer {
        private final String name;
        private final List<Event> events = new ArrayList<>();

        TestFollower(String name) {
            this.name = name;
        }

        @Override
        public void update(Event event) {
            events.add(event);
        }
    }
}

File: kotlin/social-media-photo-app/src/test/kotlin/com/rbleggi/socialmedia/SocialMediaTest.kt
package com.rbleggi.socialmedia

import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class SocialMediaTest {
    @Test
    fun testObserverPattern() {
        val user = User("Alice")
        var eventCount = 0

        val testObserver = object : Observer {
            override fun update(event: Event) {
                eventCount++
            }
        }

        user.registerObserver(testObserver)

        val photo = Photo("1", "http://photos.com/1")
        user.publishPhoto(photo)
        user.tagUser(photo, "Bob")
        user.commentPhoto(photo, Comment("Charlie", "Great!"))

        assertEquals(3, eventCount)
    }

    @Test
    fun testPhotoPublishedEvent() {
        val user = User("Alice")
        var publishedPhoto: Photo? = null

        val observer = object : Observer {
            override fun update(event: Event) {
                if (event is PhotoPublished) {
                    publishedPhoto = event.photo
                }
            }
        }

        user.registerObserver(observer)
        val photo = Photo("1", "http://photos.com/1", listOf("vacation", "beach"))
        user.publishPhoto(photo)

        assertEquals("1", publishedPhoto?.id)
        assertEquals("http://photos.com/1", publishedPhoto?.url)
        assertEquals(2, publishedPhoto?.tags?.size)
    }

    @Test
    fun testUserTaggedEvent() {
        val user = User("Alice")
        var taggedUser: String? = null
        var taggedPhoto: Photo? = null

        val observer = object : Observer {
            override fun update(event: Event) {
                if (event is UserTagged) {
                    taggedUser = event.taggedUser
                    taggedPhoto = event.photo
                }
            }
        }

        user.registerObserver(observer)
        val photo = Photo("1", "http://photos.com/1")
        user.tagUser(photo, "Bob")

        assertEquals("Bob", taggedUser)
        assertEquals("1", taggedPhoto?.id)
    }

    @Test
    fun testPhotoCommentedEvent() {
        val user = User("Alice")
        var receivedComment: Comment? = null
        var commentedPhoto: Photo? = null

        val observer = object : Observer {
            override fun update(event: Event) {
                if (event is PhotoCommented) {
                    receivedComment = event.comment
                    commentedPhoto = event.photo
                }
            }
        }

        user.registerObserver(observer)
        val photo = Photo("1", "http://photos.com/1")
        val comment = Comment("Bob", "Amazing shot!")
        user.commentPhoto(photo, comment)

        assertEquals("Bob", receivedComment?.user)
        assertEquals("Amazing shot!", receivedComment?.text)
        assertEquals("1", commentedPhoto?.id)
    }

    @Test
    fun testMultipleObservers() {
        val user = User("Alice")
        var observer1Events = 0
        var observer2Events = 0

        val observer1 = object : Observer {
            override fun update(event: Event) {
                observer1Events++
            }
        }

        val observer2 = object : Observer {
            override fun update(event: Event) {
                observer2Events++
            }
        }

        user.registerObserver(observer1)
        user.registerObserver(observer2)

        val photo = Photo("1", "http://photos.com/1")
        user.publishPhoto(photo)

        assertEquals(1, observer1Events)
        assertEquals(1, observer2Events)
    }

    @Test
    fun testRemoveObserver() {
        val user = User("Alice")
        var eventCount = 0

        val observer = object : Observer {
            override fun update(event: Event) {
                eventCount++
            }
        }

        user.registerObserver(observer)
        val photo1 = Photo("1", "http://photos.com/1")
        user.publishPhoto(photo1)

        assertEquals(1, eventCount)

        user.removeObserver(observer)
        val photo2 = Photo("2", "http://photos.com/2")
        user.publishPhoto(photo2)

        assertEquals(1, eventCount)
    }

    @Test
    fun testNoObservers() {
        val user = User("Alice")
        val photo = Photo("1", "http://photos.com/1")

        user.publishPhoto(photo)
        user.tagUser(photo, "Bob")
        user.commentPhoto(photo, Comment("Charlie", "Nice!"))
    }

    @Test
    fun testMultiplePhotos() {
        val user = User("Alice")
        val publishedPhotos = mutableListOf<Photo>()

        val observer = object : Observer {
            override fun update(event: Event) {
                if (event is PhotoPublished) {
                    publishedPhotos.add(event.photo)
                }
            }
        }

        user.registerObserver(observer)

        val photo1 = Photo("1", "http://photos.com/1")
        val photo2 = Photo("2", "http://photos.com/2")
        val photo3 = Photo("3", "http://photos.com/3")

        user.publishPhoto(photo1)
        user.publishPhoto(photo2)
        user.publishPhoto(photo3)

        assertEquals(3, publishedPhotos.size)
        assertEquals("1", publishedPhotos[0].id)
        assertEquals("2", publishedPhotos[1].id)
        assertEquals("3", publishedPhotos[2].id)
    }

    @Test
    fun testFollowerReceivesAllEventTypes() {
        val user = User("Alice")
        val follower = Follower("Bob")
        val events = mutableListOf<Event>()

        val trackingObserver = object : Observer {
            override fun update(event: Event) {
                events.add(event)
            }
        }

        user.registerObserver(follower)
        user.registerObserver(trackingObserver)

        val photo = Photo("1", "http://photos.com/1")
        user.publishPhoto(photo)
        user.tagUser(photo, "Charlie")
        user.commentPhoto(photo, Comment("Bob", "Great!"))

        assertEquals(3, events.size)
        assertTrue(events[0] is PhotoPublished)
        assertTrue(events[1] is UserTagged)
        assertTrue(events[2] is PhotoCommented)
    }

    @Test
    fun testPhotoWithTags() {
        val photo = Photo("1", "http://photos.com/1", listOf("sunset", "beach", "vacation"))

        assertEquals("1", photo.id)
        assertEquals("http://photos.com/1", photo.url)
        assertEquals(3, photo.tags.size)
        assertTrue(photo.tags.contains("sunset"))
        assertTrue(photo.tags.contains("beach"))
        assertTrue(photo.tags.contains("vacation"))
    }

    @Test
    fun testPhotoWithoutTags() {
        val photo = Photo("1", "http://photos.com/1")

        assertEquals("1", photo.id)
        assertEquals("http://photos.com/1", photo.url)
        assertTrue(photo.tags.isEmpty())
    }

    @Test
    fun testComment() {
        val comment = Comment("Alice", "Beautiful photo!")

        assertEquals("Alice", comment.user)
        assertEquals("Beautiful photo!", comment.text)
    }
}
